### ISSUE TITLE
Duplicate Schedule/Hearing Tasks

### ISSUE THREAD

**Issue by lomaxap** [department-of-veterans-affairs/caseflow#10630]
https://github.com/department-of-veterans-affairs/appeals-support/issues/5128

**Root Cause**
When a ColocatedTask is closed, [it can create a new Schedule/Hearing Task](https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/models/tasks/colocated_task.rb#L136). There is a known issue (#10337) where duplicate ColocatedTasks are being created, which produces duplicate Schedule/HearingTask. Once that is issue is resolved

- [ ] Find duplicate active Schedule/HearingTasks and cancel one (see also #10854)

**Comment by pshahVA** [comment]
Thanks Andrew. So how do I proceed?

**Comment by Chingujo** [comment]
This duplicate task is related to: 

#10522 Null dispostions not being created

#10854 Duplicate hearing tasks for legacy

**Comment by lomaxap** [comment]
depends on #10337

**Comment by lomaxap** [comment]
```
appeals_with_dupe_hearing_tasks = LegacyAppeal.where("\
id in (SELECT \
appeal_id \
FROM tasks \
WHERE appeal_id=legacy_appeals.id AND appeal_type='LegacyAppeal' \
AND status NOT IN ('completed', 'cancelled') AND type='HearingTask' \
GROUP BY appeal_id \
HAVING count(appeal_id) > 1)")

appeals_with_dupe_hearing_tasks.each do |d|
	tasks = d.tasks.order(:created_at).where(type: 'HearingTask').where.not(status: ['completed', 'cancelled'])
	task_to_close = tasks[0]
	tasks.each do |t|
		# HearingTasks with 2 or more children have a hearing scheduled (DispositionTask and ScheduleHearingTask)
	  # HearingTasks with 1 are waiting to be scheduled
		tasks_to_close = t if t.children.count < task_to_close.children.count
	end

	task_to_close.update(status: Constants.TASK_STATUSES.cancelled, instructions: ['Duplicate task closed by admin.'])
end
```

**Comment by tomas-nava** [comment]
- Since running this will affect more than 500 appeals, I think we should make it a rake task that logs its progress and the changes it makes.
- I get the exact same results whether I include the `appeal_id=legacy_appeals.id` condition in the query or not...and I'm not sure what it's supposed to do? It works in the context of the ActiveRecord query but not as straight SQL.
- I think there are some edge cases that might need special consideration. In the set of `appeals_with_dupe_hearing_tasks` there's:
  - 1 appeal (_331449_) that has 2 `HearingTask`s, neither with a hearing, each with 1 `ScheduleHearingTask`, one of which is cancelled, the other, assigned. This script wouldn't distinguish between the two, though I think it should choose the one that has the child that's assigned.
  - 7 appeals with multiple `HearingTask` children, one of which has only a `DispositionTask` child; and 2 of those `DispositionTask`s have `NoShowHearingTask` children of their own that've been put on hold. (See LegacyAppeal _614009_ for example)
  - 46 appeals that have a child `HearingTask` with no children. (like LegacyAppeal _8629_; probably not a problem)

**Comment by lomaxap** [comment]
> I think we should make it a rake task that logs its progress and the changes it makes

👍 
> I get the exact same results whether I include the `appeal_id=legacy_appeals.id` condition in the query or not

Ah, right. I was writing it as if it were a join ON statement. removing.
>   * 1 appeal (_331449_) that has 2 `HearingTask`s, neither with a hearing, each with 1 `ScheduleHearingTask`, one of which is cancelled, the other, assigned. This script wouldn't distinguish between the two, though I think it should choose the one that has the child that's assigned.
>   * 7 appeals with multiple `HearingTask` children, one of which has only a `DispositionTask` child; and 2 of those `DispositionTask`s have `NoShowHearingTask` children of their own that've been put on hold. (See LegacyAppeal _614009_ for example)
>   * 46 appeals that have a child `HearingTask` with no children. (like LegacyAppeal _8629_; probably not a problem)

Got it. So there are some duplicates in there that weren't created by closed `ColocatedTask`s. The Safest bet may be to only automatically close ones that we know were duplicated there, and look at the others separately. 

```ruby
def unchanged_colocated_dupe?(hearing_tasks)
  hearing_tasks.count == 2 && hearing_tasks.all? do |t|
     open_descendants = t.descendants.select { |d| d.open? }
     open_descendants.pluck(:type).sort == ['HearingTask', 'ScheduleHearingTask']
  end
end

def one_hearing_task_was_advanced?(hearing_tasks)
  hearing_tasks.count == 2 && hearing_tasks.select { |t| t.hearing.present? }.count == 1
end

def are_workable_tasks?(hearing_tasks)
  unchanged_colocated_dupe?(hearing_tasks) || one_hearing_task_was_advanced?(hearing_tasks)
end

# > 465 / 505 appeals 

appeals_with_dupe_hearing_tasks.each do |appeal|
 hearing_tasks = appeal.tasks.open.order(:created_at).where(type: 'HearingTask')
 next if !are_workable_tasks(hearing_tasks)

 task_to_close = nil
 if one_hearing_task_was_advanced?(hearing_tasks)
   task_to_close = hearing_tasks.find { |t| t.hearing.nil? }
 else
   task_to_close = hearing_tasks[0]
 end

  task_to_close.update(status: Constants.TASK_STATUSES.cancelled, instructions: ['Duplicate task closed by admin.'])
end
```

**Comment by lomaxap** [comment]
Tomas and I paired to close 493 duplicate hearing tasks. I still need to go through and assess the 12 appeals that could not be bucketed with the script below

final script: 
```ruby
def unchanged_colocated_dupe?(hearing_tasks)
  hearing_tasks.count == 2 && hearing_tasks.all? do |t|
    open_descendants = t.descendants.select(&:open?)
    open_descendants.pluck(:type).sort == %w[HearingTask ScheduleHearingTask]
  end
end

def one_hearing_task_was_advanced?(hearing_tasks)
  hearing_tasks.count == 2 && hearing_tasks.select { |t| t.hearing.present? }.count == 1
end

def one_hearing_task_has_no_children?(hearing_tasks)
  hearing_tasks.count == 2 && hearing_tasks.select { |t| t.children.count == 0 }.count == 1
end

appeals_with_dupe_hearing_tasks = LegacyAppeal.where("\
id in (SELECT \
appeal_id \
FROM tasks \
WHERE appeal_type='LegacyAppeal' \
AND status NOT IN ('completed', 'cancelled') AND type='HearingTask' \
GROUP BY appeal_id \
HAVING count(appeal_id) > 1)").includes(:tasks)


tasks_to_close = []
unclosable_appeals = []
appeals_with_dupe_hearing_tasks.each do |appeal|
  hearing_tasks = appeal.tasks.open.order(:created_at).where(type: "HearingTask")

  task_to_close = if unchanged_colocated_dupe?(hearing_tasks)
                    hearing_tasks[0]
                  elsif one_hearing_task_was_advanced?(hearing_tasks)
                    hearing_tasks.find { |t| t.hearing.nil? }
                  elsif one_hearing_task_has_no_children?(hearing_tasks)
                    hearing_tasks.find { |t| t.children.count == 0 }
                  end

  if task_to_close
    tasks_to_close << task_to_close
  else
    unclosable_appeals << appeal
  end
end

tasks_to_close.each do |task_to_close|
  instructions = (task_to_close.instructions || []) << "Duplicate task closed by admin."
  task_to_close.update(status: Constants.TASK_STATUSES.cancelled, instructions: instructions)
  puts "cancelled task #{task_to_close.id}"
end

puts "appeals with unclosable #{unclosable_appeals.pluck(:id).join(', ')}"
```

**Comment by lomaxap** [comment]
Turns out, `ColocatedTask`s are duplicated [intentionally](https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/models/task.rb#L400-L406) with another task being created as a child and assigned to a user with the original task as a parent. When the user completed that child task,  the parent task was also completed. This resulted in `ScheduleHearingTask.create` being called twice: once for the child and then again for the parent. 

#11177 resolved this and will be deployed 06/24/2019. I will need to rerun above script to catch any new duplicate `HearingTask`s

**Comment by hschallhorn** [comment]
Another one:
```ruby
puts Task.find(189772).appeal.structure_render(:id, :status, :assigned_at)
LegacyAppeal 315354 [id, status, assigned_at]
├── RootTask 148643, on_hold, 2019-03-21 17:40:57 UTC
│   ├── HearingTask 189772, on_hold, 2019-05-01 12:29:03 UTC
│   │   ├── ScheduleHearingTask 189773, completed, 2019-05-01 12:29:03 UTC
│   │   └── AssignHearingDispositionTask 506299, assigned, 2019-11-06 14:25:43 UTC
│   └── HearingTask 189774, on_hold, 2019-05-01 12:29:04 UTC
│       └── ScheduleHearingTask 189775, assigned, 2019-05-01 12:29:04 UTC
└── ScheduleHearingColocatedTask 166893, completed, 2019-04-07 15:57:33 UTC
    └── ScheduleHearingColocatedTask 166894, completed, 2019-04-07 15:57:33 UTC
```

**Comment by lowellrex** [comment]
Closed 487 `ScheduleHearingTask`s that should have been closed by the cancellation of their parent task in the earlier cleanup. 

```ruby
  rails c> cancelled_hearing_tasks_10630 = HearingTask.where.not(closed_at: nil).where(instructions: ["Duplicate task closed by admin."])
  rails c> cancelled_hearing_tasks_10630.count
  # 487
  rails c> schedule_tasks_to_cancel = ScheduleHearingTask.open.where(parent_id: cancelled_hearing_tasks_10630.pluck(:id))
  rails c> schedule_tasks_to_cancel.count
  # 239
  # Matches our SQL query: https://caseflow-looker.va.gov/sql/gcshbwk7xkmwds
  rails c> schedule_tasks_to_cancel.pluck(:id).sort
  # [170327, 170331, 175043, 175047, 175054, 175056, 175063, 176757, 176760, 186433, 187881, 188226, 188920, 188998, 189027, 189584, 189900, 189911, 189915, 190969, 192123, 192127, 192158, 194637, 194862, 194866, 196435, 198096, 198502, 198573, 198580, 199170, 199178, 199194, 199227, 199440, 201493, 201515, 201536, 201545, 201603, 201629, 201651, 201657, 201664, 201670, 201694, 201703, 201740, 201744, 201748, 201760, 201875, 201898, 201903, 202292, 202359, 202824, 202832, 202849, 202862, 202866, 202894, 202904, 202908, 202920, 202934, 202942, 202950, 202955, 202959, 202977, 202979, 202983, 202992, 203000, 204212, 204451, 204460, 204674, 205785, 206236, 206243, 206562, 206585, 206595, 206599, 206615, 206626, 206649, 206659, 206663, 206668, 206673, 206718, 208676, 210370, 210547, 210576, 210581, 210600, 210605, 210609, 210706, 210946, 210954, 210964, 211030, 211671, 211680, 211684, 211693, 211754, 212549, 212555, 212559, 212639, 212643, 212663, 212670, 212874, 214087, 214096, 216030, 216906, 217017, 217023, 217028, 217040, 217051, 217107, 217214, 217511, 217730, 217829, 217887, 217929, 217938, 217949, 217968, 217973, 218386, 219352, 220065, 220680, 220688, 224242, 224510, 224522, 224559, 224569, 224731, 224738, 224955, 225243, 225497, 226324, 226330, 226334, 226489, 226513, 227215, 229561, 229565, 229581, 229600, 231068, 231075, 231082, 232066, 232337, 232431, 232455, 232464, 232481, 232488, 232763, 232783, 232792, 232804, 233473, 233994, 234030, 234043, 234238, 234690, 235435, 235439, 235448, 235454, 235528, 235853, 235872, 235916, 235924, 235928, 236009, 236540, 236923, 237038, 238277, 238287, 238311, 238324, 238328, 238338, 238389, 238405, 238411, 238415, 238427, 238433, 238440, 238445, 238456, 238465, 240358, 240362, 240375, 240474, 240786, 240828, 241348, 242051, 242057, 242086, 242419, 242427, 242472, 242480, 242485, 242489, 243534, 243542, 243888, 244942, 246097, 247572, 248665]
  # Let's test one
  rails c> test_task = Task.find(170327)
  rails c> explanation = "Closing child tasks left open after closing parent tasks as part of Github ticket #10630"
  rails c> test_task.update!(status: Constants.TASK_STATUSES.cancelled, instructions: [test_task.instructions, explanation].flatten)
  # Worked. Let's refetch the set of tasks to work on and run this for all 'em.
  rails c> schedule_tasks_to_cancel = ScheduleHearingTask.open.where(parent_id: cancelled_hearing_tasks_10630.pluck(:id))
  rails c> schedule_tasks_to_cancel.count
  # 238
  # One fewer because we are no longer picking up Task ID 170327.
 rails c> schedule_tasks_to_cancel.each { |t| t.update_columns(status: Constants.TASK_STATUSES.cancelled, instructions: [t.instructions, explanation].flatten) }
  # Success. Are any left?
  rails c> ScheduleHearingTask.open.where(parent_id: cancelled_hearing_tasks_10630.pluck(:id)).count
  # 0
```

Working on investigating `ScheduleHearingTask`s that were not cancelled in the initial cleanup but that were worked on and were not in an "open" state this morning.

**Comment by lowellrex** [comment]
Closed a re-opened `RootTask` and `TrackVeteranTask` previously missed:

```ruby
  rails c> task_id = 283885
  rails c> RootTask.find(task_id).appeal.location_code
  # "99"
  rails c> puts RootTask.find(task_id).appeal.structure_render(:id, :status, :created_at, :updated_at, :closed_at)
  # Just need to close out the parent RootTask here.
  rails c> RootTask.find(task_id).update!(status: Constants.TASK_STATUSES.cancelled)
```

And created a ticket to address active `HearingTask`s related to re-opened `RootTask`s: #12678.
